chameleon/views.py

```diff
@@ -48,38 +47,46 @@ def horizon_sso_login(request):
-            # send users to a page that lets them know when their approved project begins
-            logger.debug('User: ' + request.user.username + ' is attempting to log in to Horizon, approved project found in TAS but is not yet active, sending to informational page')
-            return render(request, 'sso/chameleon_project_approved.html', {'active_date': datetime.strftime(earliest_start_date,'%B %d, %Y at %I:%M %p'),})
-        else:
+            # Error accessing Keystone, if this is the result of a form post, let's not propogate form posts any further
+            if request.method == 'POST':
+                logger.error("Encountered an error while accessing keystone to check for user projects, lets send to unavailable page")
+                return HttpResponseRedirect('/sso/horizon/unavailable')
+        if active_ks_projects_found: 
             '''
-            If we're here the user has an active project in TAS but is either missing a token or the active projects in TAS don't exist in Keystone
-            Failing to retrieve an unscoped token or projects from KS when the user has an active project in TAS may happen if the user's account is in a transient state so
-            We give the user a chance to retrieve a token
+            If we're here, then users have an active project and an unscoped token, so we proceed to log them in to Horizon
             '''
-            if not unscoped_token:
-                return manual_ks_login(request)
+            try:
+                return redirect_to_horizon(request)
+            except Exception as e:
+                logger.error(e)
+                return HttpResponseRedirect('/sso/horizon/unavailable')
+        else:
+            # no projects found in Keystone and we also did a sync at "user_has_active_ks_project(ks, request.user)", so check with TAS for approved projects
+            chameleon_projects = get_user_chameleon_projects(request.user)
+            # there are no approved projects, let's check for pending projects and their start-date
+            if chameleon_projects.get('approved_projects'):
+                earliest_start_date = None
+                for p in chameleon_projects.get('approved_projects'):
+                    for a in p.allocations:
+                        curr_alloc_date = datetime.strptime(a.start,'%Y-%m-%dT%H:%M:%SZ')
+                        if earliest_start_date is None or curr_alloc_date < earliest_start_date:
+                            earliest_start_date = curr_alloc_date
+                # send users to a page that lets them know when their approved project begins
+                logger.debug('User: ' + request.user.username + ' is attempting to log in to Horizon, approved project found in TAS but is not yet active, sending to informational page')
+                return render(request, 'sso/chameleon_project_approved.html', {'active_date': datetime.strftime(earliest_start_date,'%B %d, %Y at %I:%M %p'),})
```

Comment:
What does this do if `earliest_start_date` is `None`, which can happen if there are no allocations?